store/postgres/src/store.rs

```diff
@@ -665,17 +704,116 @@ impl Store {
+            // closure failed.
+            let conn = store.get_conn()?;
+
+            // It is possible time has passed while establishing a connection.
+            // Time to check for cancel.
+            cancel_handle.check_cancel()?;
+
+            f(conn, &cancel_handle)
+        })
+        .await;
+
+        drop(cancel_guard);
+
+        // Finding cancel isn't technically unreachable, since there is nothing
+        // stopping the supplied closure from returning Canceled even if the
+        // supplied handle wasn't canceled. That would be very unexpected, the
+        // doc comment for this function says we will panic in this scenario.
+        match result {
+            Ok(t) => Ok(t),
+            Err(CancelableError::Error(e)) => Err(e),
+            Err(CancelableError::Cancel) => panic!("The closure supplied to with_entity_conn must not return Err(Canceled) unless the supplied token was canceled."),
+        }
+    }
+
+    async fn with_entity_conn<T: Send + 'static>(
+        &self,
+        subgraph: &SubgraphDeploymentId,
+        f: impl 'static + Send + FnOnce(&e::Connection, &CancelHandle) -> Result<T, CancelableError>,
+    ) -> Result<T, Error> {
+        let store = self.cheap_clone();
+        // Unfortunate clone in order to pass a 'static closure to with_conn.
+        let subgraph = subgraph.clone();
+
+        // Duplicated logic: No need to make re-usable when the
+        // other end will go away.
+        // See also 220c1ae9-3e8a-42d3-bcc5-b1244a69b8a9
+        let start = Instant::now();
+        let registry = self.registry.cheap_clone();
+
+        self.with_conn(move |conn, cancel_handle| {
+            registry
+                .global_counter(format!("{}_get_entity_conn_secs", &subgraph))
+                .map_err(Into::<Error>::into)?
+                .inc_by(start.elapsed().as_secs_f64());
+
+            cancel_handle.check_cancel()?;
+            let storage = store
+                .storage(&conn, &subgraph)
+                .map_err(Into::<Error>::into)?;
+            cancel_handle.check_cancel()?;
+            let metadata = store
+                .storage(&conn, &*SUBGRAPHS_ID)
+                .map_err(Into::<Error>::into)?;
+            cancel_handle.check_cancel()?;
+            let conn = e::Connection::new(conn, storage, metadata);
+
+            f(&conn, cancel_handle)
+        })
+        .await
+    }
+
     fn get_conn(&self) -> Result<PooledConnection<ConnectionManager<PgConnection>>, Error> {
         let start_time = Instant::now();
         let conn = self.conn.get();
         let wait = start_time.elapsed();
         if wait > Duration::from_millis(10) {
-            warn!(self.logger, "Possible contention in DB connection pool";
-                               "wait_ms" => wait.as_millis())
+            warn!(&self.logger, "Possible contention in DB connection pool";
+                                "wait_ms" => wait.as_millis())
```

Comment:
I'd be hugely in favor of just getting rid of this log statement. I don't think we get much utility out of it anymore.